src/main/java/org/akhq/modules/OidcUserDetailsMapper.java

```diff
@@ -0,0 +1,115 @@
+ */
+@Singleton
+@Replaces(DefaultOpenIdUserDetailsMapper.class)
+@Requires(property = "akhq.security.oidc.enabled", value = StringUtils.TRUE)
+public class OidcUserDetailsMapper extends DefaultOpenIdUserDetailsMapper {
+    private final UserGroupUtils userGroupUtils;
+    private final Oidc oidc;
+
+    @Value("${akhq.security.default-group}")
+    private String defaultGroups;
+
+    /**
+     * Default constructor.
+     *
+     * @param openIdAdditionalClaimsConfiguration The additional claims configuration
+     * @param oidc the OIDC configuration
+     * @param userGroupUtils the utils class to translate user groups
+     */
+    public OidcUserDetailsMapper(
+            OpenIdAdditionalClaimsConfiguration openIdAdditionalClaimsConfiguration,
+            Oidc oidc,
+            UserGroupUtils userGroupUtils
+    ) {
+        super(openIdAdditionalClaimsConfiguration);
+        this.oidc = oidc;
+        this.userGroupUtils = userGroupUtils;
+    }
+    /**
+     * Tries to read the username from the configured username field.
+     *
+     * @param providerName The OpenID provider name
+     * @param tokenResponse The token response
+     * @param openIdClaims The OpenID claims
+     * @return The username to set in the {@link UserDetails}
+     */
+    @Override
+    protected String getUsername(String providerName, OpenIdTokenResponse tokenResponse, OpenIdClaims openIdClaims) {
+        Oidc.Provider provider = oidc.getProvider(providerName);
+        return Objects.toString(openIdClaims.get(provider.getUsernameField()));
+    }
+
+    /**
+     * Tries to read roles from the configured roles field and translates them using {@link UserGroupUtils}.
+     * If the configured field cannot be found or isn't some kind of collection, it will use the default group.
+     *
+     * @param providerName The OpenID provider name
+     * @param tokenResponse The token response
+     * @param openIdClaims The OpenID claims
+     * @return The roles to set in the {@link UserDetails}
+     */
+    @Override
+    protected List<String> getRoles(String providerName, OpenIdTokenResponse tokenResponse, OpenIdClaims openIdClaims) {
+        return userGroupUtils.getUserRoles(getOidcRoles(providerName, openIdClaims));
+    }
+    /**
+     * Adds the configured attributes for the user roles to the user attributes.
+     *
+     * @param providerName The OpenID provider name
+     * @param tokenResponse The token response
+     * @param openIdClaims The OpenID claims
+     * @return The attributes to set in the {@link UserDetails}
+     */
+    @Override
+    protected Map<String, Object> buildAttributes(String providerName, OpenIdTokenResponse tokenResponse, OpenIdClaims openIdClaims) {
+        Map<String, Object> attributes = super.buildAttributes(providerName, tokenResponse, openIdClaims);
+        userGroupUtils.getUserAttributes(getOidcRoles(providerName, openIdClaims)).forEach(attributes::put);
+        attributes.remove("roles");
```

Comment:
should this be the roles field?

Reply:
I was using the groups claim from Azur AD and It output to many AD/Office 365 groups for the cookie to store in the browser. Dropping the roles field cleaned up that space.

Reply:
I‘m intentionally dropping a field called „roles“ because that is where Micronaut inserts the resolved roles. It will not do that if there is an attribute called „roles“.

For the other problem, rather than just copying everything over from the provider JWT we should probably add a whitelist.

Reply:
To be honest, it's this part that hurt me, storing the roles in the JWT cookie.

In a future release, I will need to use another logic than role for akhq, people want to have role per cluster, that lead to incompatibility with the micronaut role.

I would have prefer to have only username in the JWT token, and server will add the role on the login part.

This will allow to change the role on server side without any impact on the cookie part and we won't have anymore the cookie size issue.

@jniebuhr what do you think about that ?

Reply:
If we just Store the username in the cookie there won’t be a way to get the roles for the user, other than caching it in-memory or doing the OIDC flow all over again. We could store all the unresolved roles that had matches in the cookie, that should be smaller.

Reply:
In another project, I use this to handle custom mapping : 
```java
import io.micronaut.security.oauth2.endpoint.token.response.DefaultOpenIdUserDetailsMapper;

@Singleton
@Replaces(DefaultOpenIdUserDetailsMapper.class)
public class GlobalOpenIdUserDetailsMapper extends DefaultOpenIdUserDetailsMapper {

    public GlobalOpenIdUserDetailsMapper(OpenIdAdditionalClaimsConfiguration openIdAdditionalClaimsConfiguration) {
        super(openIdAdditionalClaimsConfiguration);
    }

    @Override
    @Nonnull
    public UserDetails createUserDetails(String providerName, OpenIdTokenResponse tokenResponse, OpenIdClaims openIdClaims) {
        // user is here : openIdClaims.getEmail();
        // mapping username using configuration files 
        //  return new UserDetails();
    }
]
```
It works very well and nothing except the user is store on the cookie. 
The oauth flow is only done 1 time also.

Reply:
Do you have a username to roles mapping in the Config then?

Reply:
Yes but in this project I have a database to fetch the roles.

That lead to a question, where is the role in this actual implementation ?

Reply:
The roles get passed in the provider token as a claim, I then map them to AKHQ roles. Those get stored in the JWT and can be accessed on every request.

Reply:
So you have some openid provider where you can customize the claim content ? (add specific fields for an application) ?  

Most of users will use a generic openId provider (Google, Microsoft, ...) and will not be able to customize the claims (IMO)

Reply:
You can’t always customize which claims are there but you can usually select them well enough. If the provider has no understanding about roles you’ll need to use a default role anyway. In our case we have full control over claims, yes.

Reply:
Maybe we can add a default role for OIDC authenticated users

Reply:
what I have done in spring before but I believe it would be equivalent to replace the bean UserDetails and override getRoles to map the group "role" form the IDP to the RBAC roles that have been defined in the configuration. It will help keep the JWT token smaller. https://github.com/micronaut-projects/micronaut-security/blob/39eef7d3d3cfe6a25ba957026d725a78b769cbb8/security/src/main/java/io/micronaut/security/authentication/UserDetails.java

Reply:
> Most of users will use a generic openId provider (Google, Microsoft, ...) and will not be able to customize the claims (IMO)

to support this use-case there will probably need to be username to group "role" mapping in the future and then the UserDetails mapper can output the roles that are referenced on the endpoint annotation

Reply:
Agree with that, but I also think it's a must have for OpenId.
Most of users will use a public / not customizable providers so we need to allow user to have a default role but also have a special mapping for username > roles for admin for example :) 
I will hope to have a better mapping (group based) but I don't see a better way for now :cry: